### ISSUE TITLE
[common/http] HttpClientTestingBackend with async interceptor

### ISSUE THREAD

**Issue by dherges** [angular/angular#24059]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

```ts
@Injectable()
export class AsyncInterceptor implements HttpInterceptor {

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {

    return Observable.fromPromise(Promise.resolve('foo'))
      .pipe(
        switchMap((value: string) => {
          return next.handle(req.clone({ setHeaders: Foo: value }));
        }
      );
  }
}
```

```ts
  it(`should set Foo header`, async(inject([HttpTestingController, HttpClient],
    (backend: HttpTestingController, client: HttpClient) => {
      client.get('/').subscribe((v) => { alert(v); });

      backend.expectOne((req: HttpRequest<any>) => {
                debugger; // <-- not being run
                return req.headers['Foo'] === 'foo';
        }, `expected 'Authorization: Bearer foo'`).flush('foobar');

     backend.verify(); // <-- will fail the test
  })));
````

## Expected behavior

The issue here is that `expectOne()` calls into [`HttpClienTestingBackend#match()`](https://github.com/angular/angular/blob/master/packages/common/http/testing/src/backend.ts#L54) is done before [`HttpClientTestingBackend#handle()`](https://github.com/angular/angular/blob/master/packages/common/http/testing/src/backend.ts#L41)

## Minimal reproduction of the problem with instructions

Stackblitz Repro:

https://stackblitz.com/edit/angular-testing-template-cz2iym

## What is the motivation / use case for changing the behavior?

testing with `HttpClient` and an async interceptor in the chain

## Environment

<pre><code>
Angular version: 6.0.0
</code></pre>

**Comment by dherges** [comment]
In the above example,. the issue goes down to [promises being resolved async](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#Return_value). I'd probably could get to work around by moving the `backend.expectOne()` inside a `setTimeout()` block?!

**Comment by alxhub** [comment]
@dherges has the right idea. It's up to you to ensure your HTTP request has resolved fully before calling `verify`. This includes waiting for any async logic in your interceptors.

**Comment by dherges** [comment]
But it's still not very nice. Even if I wrap the test code in `fakeAsync` I have to use a `setTimeout` to get the expected order of promise resolution.

---

**TL;DR** `setTimeout` ftw!
